### PR TITLE
Fuyu processing: handle coordinates

### PR DESCRIPTION
**Pull request by pcuenca**
Replaces https://github.com/huggingface/transformers/pull/27083

### CODE REVIEW

File: src/transformers/models/fuyu/processing_fuyu.py [hunk 1]
```diff
@@ -290,39 +285,27 @@ def _tokenize_prompts_with_image_and_batch(
...
-    cropped_coords = (
-        self._clamp_coords(original_coords, min_value=self.crop_left, max_value=self.crop_right) - self.crop_left
-    )
-    # apply scale
-    scaled_coords = self._scale_coords(cropped_coords, scale=self.scaled_w / self.original_w)
-    # apply pad
-    return scaled_coords + self.padding_left
+# Simplified assuming self.crop_left = self.padding_left = 0
+def original_to_transformed_w_coords(original_coords, scale_w):
+    return np.round(original_coords * scale_w).astype(np.int32)
```
Comment (pcuenca) [inline src/transformers/models/fuyu/processing_fuyu.py:295]: I haven't seen crop_top, crop_left, padding_left and padding_top different to zero in the original code (see https://huggingface.co/adept-hf-collab/adept-mm/blob/736c6b570b2a9c0367a3266746fd1f53cfff0a2b/mm-inference-for-hf/multimodal/data/image_utils.py#L65).

File: src/transformers/models/fuyu/processing_fuyu.py [hunk 2]
```diff
@@ -591,6 +575,111 @@ def __call__(
...
+            raise ValueError("Make sure that you pass in as many target sizes as output sequences")
+        if target_sizes.shape[1] != 2:
+            raise ValueError("Each element of target_sizes must contain the size (h, w) of each image of the batch")
+
+        results = None
+        for seq, size in zip(outputs, target_sizes):
+            seq = tokens_to_boxes(seq, size)
+            seq = tokens_to_points(seq, size)
+            seq = seq[None, :]
+            # TODO: what if sequence lengths vary?
```
Comment (pcuenca) [inline src/transformers/models/fuyu/processing_fuyu.py:678]: I haven't tested this yet, but if we use batch inference and there are different number of encoded boxes or points in the sequences, the tensor shapes will be different after these transformations. How can we deal with this case?

**Comment by pcuenca** [comment]
@amyeroberts @molbap could you please take a first look?

There's a TODO about batch handling when coordinates are translated that I'm not sure how to resolve. Other than that, I'll check for potentially unused coordinate handling code that was brought over from the original repo.

File: src/transformers/models/fuyu/processing_fuyu.py (hunk 2 above)
Comment (amyeroberts) [reply src/transformers/models/fuyu/processing_fuyu.py:678]: Not sure what's best. I can see two ways: either we have a a way of padding the sequences to the max length then concatenating, or we return a list of variable length tensors. 

If I've understood the method correctly, this is postprocessing the model outputs and not producing the raw outputs which would be fed back to the model when generating the next token. In this case, do we need to return a batched tensor?

**Review by amyeroberts — APPROVED** [review]
Very nice! 🔥

File: src/transformers/models/fuyu/processing_fuyu.py (hunk 2 above)
Comment (molbap) [reply src/transformers/models/fuyu/processing_fuyu.py:678]: I think returning a list of variable lengths tensors is fine, there's few things after generation that'd leverage a padded version of it. Just something like 
```
results = []
for seq, size in zip(outputs, target_sizes):
    seq = tokens_to_boxes(seq, size)
    seq = tokens_to_points(seq, size)
    results.append(seq)
```

File: src/transformers/models/fuyu/processing_fuyu.py [hunk 3]
```diff
@@ -591,6 +575,111 @@ def __call__(
...
+
+                tokens = torch.cat([tokens[:start], replacement, tokens[end+1:]], 0)
+            return tokens
+
+        if target_sizes is None:
+            target_sizes = ((self.image_processor.size["height"], self.image_processor.size["width"]),) * len(outputs)
+
+        if len(outputs) != len(target_sizes):
+            raise ValueError("Make sure that you pass in as many target sizes as output sequences")
+        if target_sizes.shape[1] != 2:
```
Comment (molbap) [inline src/transformers/models/fuyu/processing_fuyu.py:670]: here if target_sizes is None, this will error out as .shape is not a tuple method. If it is a tuple, this check is not necessary as we set the sizes manually

Comment (molbap) [reply src/transformers/models/fuyu/processing_fuyu.py:670]: cc @pcuenca

File: src/transformers/models/fuyu/processing_fuyu.py (hunk 2 above)
Comment (pcuenca) [reply src/transformers/models/fuyu/processing_fuyu.py]: Yes, you are both right, thanks! I've checked in other post-processors that there's no need to keep the same input type if it's not necessary.

File: src/transformers/models/fuyu/processing_fuyu.py (hunk 3 above)
Comment (pcuenca) [reply src/transformers/models/fuyu/processing_fuyu.py]: Fixed, thanks a lot, @molbap!
